WireAuthentication/Sources/WireAuthenticationLogic/DetermineAuthMethodUseCase.swift

```diff
@@ -17,17 +17,93 @@
 //
 
 import Foundation
+import WireAPI
 import WireAuthenticationAPI
 
 package struct DetermineAuthMethodUseCase: DetermineAuthMethodUseCaseProtocol {
 
-    package init() {}
+    private let validateEmailOrSSOCode: ValidateEmailOrSSOCodeUseCase
+    private let authenticationAPI: AuthenticationAPI
+
+    package init(validateEmailOrSSOCode: ValidateEmailOrSSOCodeUseCase, authenticationAPI: AuthenticationAPI) {
+        self.validateEmailOrSSOCode = validateEmailOrSSOCode
+        self.authenticationAPI = authenticationAPI
+    }
+
+    @MainActor
+    package func invoke(
+        emailOrSSOCode: String
+    ) async throws(DetermineAuthMethodUseCaseFailure) -> AuthenticationMethod {
+        let emailOrSSOCode = try validateEmailOrSSOCode(input: emailOrSSOCode)
+
+        switch emailOrSSOCode {
+        case let .email(email, domain):
+            do {
+                return try await determineAuthMethod(email: email, domain: domain)
+            } catch let error as DetermineAuthMethodUseCaseFailure {
+                throw error
+            } catch AuthenticationAPIError.invalidResponse {
+                throw .invalidResponse
+            } catch let error as URLError {
+                throw .urlError(error)
+            } catch {
+                throw .unknown
+            }
+        case let .ssoCode(ssoCode):
+            return .loginViaSSO(code: ssoCode)
+        }
+    }
+
+    // MARK: - Private
+
+    private func validateEmailOrSSOCode(
+        input: String
+    ) throws(DetermineAuthMethodUseCaseFailure) -> ValidatedEmailOrSSOCode {
+        do {
+            return try validateEmailOrSSOCode.invoke(input: input)
+        } catch {
+            throw .invalidEmailOrSSOCode
+        }
+    }
 
     @MainActor
-    package func invoke(emailOrSSOCode: String) async throws -> AuthenticationMethod {
-        try await Task.sleep(for: .seconds(3))
+    private func determineAuthMethod(email: String, domain: String) async throws -> AuthenticationMethod {
+        let configuration: DomainRegistrationConfiguration
+        do {
+            configuration = try await authenticationAPI.getDomainRegistration(forEmail: email)
+        } catch AuthenticationAPIError.unsupportedEndpointForAPIVersion {
+            // Fallback if the API doesn't support the getDomainRegistration endpoint
+
+            do {
+                let onPremConfig = try await authenticationAPI.getOnPremConfigURL(forDomain: domain)
+                return .onPremLogin(email: email, backendConfig: onPremConfig.configurationURL)
```

Comment:
@johnxnguyen @samwyndham Do we want to use it for clients who are still using API below v8? If so, from a UI perspective, it will show a redirect UI screen. It might be confusing for our old clients to see the redirect backend screen every time. What do you think?

Reply:
let's keep it for now and will test later